[hjuntan] iP

hjuntan opened the pull request:
# Toothless
> “It’s 100% easier to increase your time and freedom by eliminating the dumb things you do every day than to try to be 100% more productive doing more dumb things.” - Richie Norton ([source](https://quotefancy.com/quote/3802958/Richie-Norton-It-s-100-easier-to-increase-your-time-and-freedom-by-eliminating-the-dumb))

## What is Toothless? 

Toothless is a chatbot that helps free your mind of having to remember things you need to do. It's:
- text-based
- simple to learn and understand
- ~~MURD-~~ FAST to use

## Where can I find Toothless?

All you need to do is:

1. Download it from [here](https://github.com/hjuntan/ip.git)
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you :star_struck: 

And it is FREE!

## What does Toothless have?

- [X]  Managing todos
- [X]  Managing deadlines
- [X]  Managing events
- [ ]  Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Toothless {
    public static void main(String[] args) {
        new Toothless("./data/toothless.txt").start();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

hjuntan committed:
Implement skeletal version of chatbbot
Change name of chatbot from Duke to Toothless

hjuntan committed:
Toothless will now echo the command entered by the user, and exit when the user types the command "bye"

hjuntan committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested using the command "list".

hjuntan committed:
Changed loop logic of chatbot

hjuntan committed:
Added the mark feature to Task

hjuntan committed:
Added the ability to mark and unmark tasks as done

hjuntan committed:
Implement the ability to create a todo, which is similar to task.

hjuntan committed:
Implement the ability to create deadlines

hjuntan committed:
Implement the ability to create events

hjuntan committed:
Added a semi-automated testing of Toothless

hjuntan committed:
Refactor code to handle commands as enums

hjuntan committed:
Handle some errors for mark, unmark, todo, deadline, event

hjuntan committed:
Better handle errors of mark and unmark

hjuntan committed:
Handle minor error of list

hjuntan committed:
Implement to ability to delete task

hjuntan committed:
Add a data file path

hjuntan committed:
Implement a method to store tasks in a txt file

hjuntan committed:
Add ability to load task into array list from text file

hjuntan committed:
Bug fix to load tasks properly

hjuntan committed:
Clean up code

hjuntan committed:
Merge branch 'branch-Level-7'

hjuntan committed:
Change the date format of deadlines and events from String to LocalDateTime

hjuntan committed:
Fix formatting issues with date time

hjuntan committed:
Add a reminder for unmark tasks to the user during launch.

hjuntan committed:
Merge branch 'branch-Level-8'

hjuntan committed:
Add more classes to handle storage, ui, tasklist and parser

hjuntan committed:
Adjust command flow in command enum based on the additional classes

hjuntan committed:
Change command from enum to an abstract class

hjuntan committed:
Merge branch 'A-MoreOOP'

hjuntan committed:
Package commands and tasks respectively

hjuntan committed:
Merge branch 'branch-A-Packages'

hjuntan committed:
Merge branch 'add-gradle-support'

hjuntan committed:
Added unit tests for MarkCommand, Todo, Deadline and Event

hjuntan committed:
Merge branch 'branch-A-JUnit'

hjuntan committed:
Package the app as a jar file

hjuntan committed:
Merge branch 'branch-A-Jar'

hjuntan committed:
Add javadoc comments for classes in the main toothless package.

hjuntan committed:
Add javadoc comments for all command classes in the toothless comands package.

hjuntan committed:
Add javadoc comments for all task classes in the toothless tasks package.

hjuntan committed:
Merge branch 'branch-A-JavaDoc'

hjuntan committed:
Tweak the code to comply with Google Java coding standard

hjuntan committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/toothless/TaskList.java
#	src/main/java/toothless/Toothless.java
#	src/main/java/toothless/ToothlessException.java
#	src/main/java/toothless/Ui.java
#	src/main/java/toothless/commands/ByeCommand.java
#	src/main/java/toothless/commands/Command.java
#	src/main/java/toothless/commands/DeadlineCommand.java
#	src/main/java/toothless/commands/DeleteCommand.java
#	src/main/java/toothless/commands/EventCommand.java
#	src/main/java/toothless/commands/ListCommand.java
#	src/main/java/toothless/commands/MarkCommand.java
#	src/main/java/toothless/commands/TodoCommand.java
#	src/main/java/toothless/commands/UnmarkCommand.java
#	src/main/java/toothless/tasks/Deadline.java
#	src/main/java/toothless/tasks/Event.java
#	src/main/java/toothless/tasks/Task.java
#	src/main/java/toothless/tasks/Todo.java

hjuntan committed:
Implement the ability to find a task by searching for a keyword in the description

hjuntan committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/toothless/tasks/Task.java

AidenLYT commented on src/main/java/toothless/Storage.java:
```diff
@@ -0,0 +1,88 @@
+                    task = new Event(storedTask[2], storedTask[3], storedTask[4], storedTask[1].equals("1"));
+                    break;
+                default:
+                    throw new ToothlessException("File corrupted O_O. Try again later.");
```
Very good use of abstraction over here 👍🏻

AidenLYT commented on src/main/java/toothless/Ui.java:
```diff
@@ -0,0 +1,87 @@
+
+    private String splitLine = "____________________________________________________________";
+    private String chatBotName = "Toothless";
+    private String greetingString = "Hi! "+ chatBotName +" is " + chatBotName + "!\n"
```
A little syntax correction to be made 
````
suggestion - spacing between arithmetic operation: 
"Hi! " + chatBotName + " is "
````

AidenLYT commented on src/main/java/toothless/Ui.java:
```diff
@@ -0,0 +1,87 @@
+     * @param task The task to be displayed.
+     */
+    public void showTask(Task task) {
+        System.out.println(" ["+ task.getTaskIcon()+"]["+ task.getStatusIcon() + "] " + task);
```
A little syntax correction here as well 
````
suggestion:
" [" + task.getTaskIcon() + "][" + task.getStatusIcon() + "] " + task);
````

AidenLYT commented on src/main/java/toothless/Ui.java:
```diff
@@ -0,0 +1,87 @@
+     * @param index The index of the task to be displayed.
+     */
+    public void showTask(Task task, int index) {
+        System.out.format("%d. ["+ task.getTaskIcon()+"]["+ task.getStatusIcon() + "] " + task + "\n", index + 1);
```
Syntax correction here as well

AidenLYT commented on src/main/java/toothless/commands/Command.java:
```diff
@@ -0,0 +1,41 @@
+ * Commands are used to interact with the application's user interface, task list, and storage system,
+ * potentially throwing a ToothlessException to signal issues specific to command execution.
+ */
+public abstract class Command {
```
Good use of abstract class 👍🏻

AidenLYT commented on src/main/java/toothless/commands/EventCommand.java:
```diff
@@ -0,0 +1,67 @@
+        if (date2Index == -1) {
+            throw new ToothlessException("Human event no end date D:");
+        }
+
```
Great handling of different errors 👍🏻

AidenLYT commented on src/main/java/toothless/tasks/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+
+    @Override
+    public String toString() {
+        return description + " (by: "+ super.dateTimeFormat(date) + ")";
```
Minor syntax error here
````
suggestion: 
description + " (by: " + super.dateTimeFormat(date) + ")"
````

AidenLYT commented on src/main/java/toothless/tasks/Task.java:
```diff
@@ -0,0 +1,84 @@
+     * @return A string "X" if the task is done, or a space " " if it is not done.
+     */
+    public String getStatusIcon() {
+        return isDone ? "X": " "; // mark done task with X
```
Minor syntax error 
````
suggestion: 
isDone ? "X" : " ";
````

AidenLYT reviewed (commented):
LGTM 👍🏻 Just a few nits to fix. Mostly minor spacing corrections. 
Code is very easy to read and follow. However, adding comments at certain parts in the code would be better to improve readability. 
Great use of abstraction as well. Keep up the good work! 😄

hjuntan committed:
Add dependencies for javafx

hjuntan committed:
Create a simple gui window

hjuntan committed:
Implement scroll pane, and handling of user input

hjuntan committed:
Add dialog boxes and user images.

hjuntan committed:
Implement gui functionality fully.

hjuntan committed:
Merge branch 'branch-Level-10'

hjuntan committed:
Add assertions into essential parts of the code

Assert the index received when retriving or removing a task from the TaskList, should not be a negative number.
Assert the stored items are in the correct format.
Assert the parser understands the correct instructions only.

hjuntan committed:
Improve code quality

Rename remaining "duke" keywords present in the project to toothless to standardise the chatbot's name.

Refactor parsing of update and task commands in parseCommand to their respective methods to improve code readability.

Improve coding standard by adding appropreiate whitespaces.

hjuntan committed:
Merge pull request #1 from hjuntan/branch-A-Assertions

Add assertions into essential parts of the code

hjuntan committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/toothless/Storage.java

hjuntan committed:
Rename the remaining "duke" keywords left behind

hjuntan committed:
Merge pull request #2 from hjuntan/branch-A-CodeQuality

Branch a code quality

NatLeong commented on src/main/java/toothless/Parser.java:
```diff
@@ -0,0 +1,92 @@
+     * @return Command corresponding to the user's request
+     * @throws ToothlessException if the input is invalid or no Command is suitable
+     */
+    public static Command parseCommand(String input) throws ToothlessException {
```
Good  idea of grouping different commands into groups for use of same parser commands. Will take inspo!

NatLeong reviewed (commented):
Use of grouping of types of instructions

NatLeong commented on src/main/java/toothless/commands/ByeCommand.java:
```diff
@@ -0,0 +1,33 @@
+package toothless.commands;
```
Overall good use of abstraction of commands

NatLeong reviewed (commented):
Use of abstractions

NatLeong commented on src/main/java/toothless/Ui.java:
```diff
@@ -0,0 +1,86 @@
+ * and task lists, as well as reading user input.
+ */
+public class Ui {
+    private String chatBotName = "Toothless";
```
I love the idea of Toothless. So cute!

NatLeong commented on src/main/java/toothless/Ui.java:
```diff
@@ -0,0 +1,86 @@
+    /**
+     * Displays a loading message indicating that previous tasks are being loaded from the file.
+     */
+    public String showLoadingTasks() {
```
Clear responses and instructions. Useful for user guidance:)

hjuntan committed:
Add new command: schedule

Schedule command is a single word command
with no extra details and not case-sensitive.

This command allows users to easily track which
task is due or starting based on the
chronological order of each task date.

hjuntan committed:
Merge branch 'branch-BCD-Extension'

hjuntan committed:
Add background in dialog

Green color for chatbot Toothles and orange color for user
Background fits the dialog with some padding

hjuntan committed:
Hightlight errors

Display bold and colored yellow to show errors made by user

hjuntan committed:
Add images of Toothless and background

hjuntan committed:
Fix bye command

An exit command will disable the user input and send button
and after 1.5 seconds, close the application.

hjuntan committed:
Fix dialog box

Dialog box will not turn to ellipsis.

hjuntan committed:
Tweak dialog box shape

hjuntan committed:
Add blank space after each date schedule

hjuntan committed:
Merge branch 'branch-BCD-Extension' into branch-A-BetterGUI

hjuntan committed:
Merge branch 'branch-A-BetterGUI'

hjuntan committed:
Improve Toothless personality

Toothless will now talk similar to a neanderthal,
which is talking in single syllabes.
The user will be refered to as human.

hjuntan committed:
Merge branch 'branch-A-Personality'

hjuntan committed:
Handle invalid date error and write file error

Invalid dates such as Feb 30 is caught
and an error message is printed for the user.
Error of unable to write to file is caught
and printed for the user

hjuntan committed:
Handle some command format errors

Handle errors such as:
- multiple spaces where only one is expected.
- trailing/leading spaces in the command.

The error message will be relayed to the user as well.

hjuntan committed:
Add scene title "Toothless"

hjuntan committed:
Merge branch 'branch-A-BetterGUI' into branch-A-MoreErrorHandling

hjuntan committed:
Merge branch 'branch-A-MoreErrorHandling'

hjuntan committed:
Minor adjustments

hjuntan committed:
Update README.md

hjuntan committed:
Move parser methods into Parser class and update JavaDoc

hjuntan committed:
Tweak README.md

hjuntan committed:
Tweak pauseTransition delay

User are able to read the bye message in time.

hjuntan committed:
Update README.md

hjuntan committed:
Change date format and improve error comments to user

Date format changed from YYYY-MM-DDThh:mm to YYYY-MM-DD hh:mm.
Improve error messages shown to user by making them more informative,
as well as slightly add more personality into them.

hjuntan committed:
Fix minor date issues

hjuntan committed:
Update minor date bug

hjuntan committed:
Update error handling of deadlines, events and update commands

Dealines and events will better detect no task description and no date,
Update commands has more concise conditional statements.

hjuntan committed:
Adjust commands in events and deadline

Deadline and event commands now has some leeway such that
description or date can appear together with / indicators.
Example: event project/from 2024-02-20 22:00/to 2024-02-20 23:00
is a valid command.

hjuntan committed:
Adjust imported packages
